### PR TITLE
materialiser ets hacks

### PR DESCRIPTION
**Pull request by tcrain**
This deals with issue
https://github.com/SyncFree/antidote/issues/83

This addresses two major issues with the materialiser scalability, without changing the functionality.

The first deals with when a new op is added.  Before we had to load the list of ops from the ets table, make a new list with the op appended, then rewrite the entire new list back into ets.  Now it just destructively updates the op in the ets table.

The second deals with workloads that have DC locality, for example where one DC would read and update a key often while another DC would rarely access this key.  In this case at the DC where the object was rarely accessed the list of operations might overflow since garbage collection was only triggered by reads.  Now the GC is forced to trigger before this happens.

There are still many things to be done with the materialiser, but these were two main issues that could be addressed without major refactoring.

### CODE REVIEW

File: src/materializer_vnode.erl [hunk 1]
```diff
@@ -281,13 +295,13 @@ internal_read(Key, Type, MinSnapshotTime, TxId, OpsCache, SnapshotCache) ->
 			ignore ->
 			    {ok, Snapshot};
 			_ ->
-			    case NewSS and IsFirst of
+			    case (NewSS and IsFirst) or ShouldGc of
```
Comment (bieniusa) [inline src/materializer_vnode.erl]: Should this be an "orelse"? and an "andalso"?

File: src/materializer_vnode.erl [hunk 2]
```diff
@@ -314,27 +328,28 @@ belongs_to_snapshot_op(SSTime, {OpDc,OpCommitTime}, OpSs) ->
 %% @doc Operation to insert a Snapshot in the cache and start
 %%      Garbage collection triggered by reads.
 -spec snapshot_insert_gc(key(), vector_orddict:vector_orddict(),
-                         cache_id(),cache_id() ) -> true.
-snapshot_insert_gc(Key, SnapshotDict, SnapshotCache, OpsCache)->
+                         cache_id(),cache_id(),boolean()) -> true.
+snapshot_insert_gc(Key, SnapshotDict, SnapshotCache, OpsCache,ShouldGc)->
     %% Should check op size here also, when run from op gc
-    case (vector_orddict:size(SnapshotDict))>=?SNAPSHOT_THRESHOLD of
+    case ((vector_orddict:size(SnapshotDict))>=?SNAPSHOT_THRESHOLD) or ShouldGc of
```
Comment (bieniusa) [inline src/materializer_vnode.erl]: Also here "orelse"?

File: src/meta_data_manager.erl [hunk 3]
```diff
@@ -74,15 +74,15 @@ init([Name]) ->
     {ok, #state{table=Table}}.
 
 handle_cast({update_meta_data, Name, NodeId, Dict}, State) ->
-    true = ets:insert(meta_data_sender:get_name(Name,?REMOTE_META_TABLE_NAME), {NodeId, Dict}),
```
Comment (bieniusa) [inline src/meta_data_manager.erl]: I guess we should keep this pattern matching here, as some form of assertion.

File: src/materializer_vnode.erl [hunk 4]
```diff
@@ -314,27 +328,28 @@ belongs_to_snapshot_op(SSTime, {OpDc,OpCommitTime}, OpSs) ->
...
+					[] ->
+					    {0, 0, []};
+					[Tuple] ->
+					    [Key,Length1,OpId1|Ops] = tuple_to_list(Tuple),
+					    {Length1, OpId1, sub_reverse(Ops,Length1,[])}
+				    end,
             {NewLength,PrunedOps}=prune_ops({Length,OpsDict}, CommitTime),
             ets:insert(SnapshotCache, {Key, PrunedSnapshots}),
-            true = ets:insert(OpsCache, {Key, {NewLength,PrunedOps}});
+            true = ets:insert(OpsCache, list_to_tuple([Key,NewLength,OpId|lists:reverse(PrunedOps) ++ lists:seq(NewLength+1,?OPS_THRESHOLD)]));
```
Comment (bieniusa) [inline src/materializer_vnode.erl]: These lists are very short, i.e. less than 10 elements, right?

**Comment by tcrain** [comment]
Ok, please review again.  I made changes so we avoid extra costs when constructing the lists and tuples during reads and GC.

The overall idea here is that if we store the operations in the ets table in a tuple per key instead of a list, then we can use a much more efficient way of adding new operations to the cache as they arrive.  We can directly update a single element in tuple in an ets table without modifying loading or modifying anything else.  Before we were making new copies of the entire list just to update a single op, then rewriting the entire list back in the table, which not only meant rewriting the entire list of operations for that key, but also invalidating the nearby pointers of the data structure as they would point to an entirely new object.

**Pull request merged**